interop-java/src/main/java/scalaz/zio/java/IO.java

```diff
@@ -0,0 +1,143 @@
+package scalaz.zio.java;
+
+import scala.runtime.BoxedUnit;
+import scalaz.zio.ExitResult;
+import scalaz.zio.Fiber;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class IO<E, A> {
+
+    private scalaz.zio.IO<E, A> delegate;
+
+    private IO(scalaz.zio.IO<E, A> delegate) {
+        this.delegate = delegate;
+    }
+
+    public <B> IO<E, B> map(Function<A, B> f) {
+        return new IO<>(delegate.map(f::apply));
+    }
+
+    public <E2, B> IO<E2, B> bimap(Function<E, E2> f, Function<A, B> g) {
+        return new IO<>(delegate.bimap(f::apply, g::apply));
+    }
+
+    public <E1 extends E, B> IO<E1, B> flatMap(Function<A, IO<E1, B>> f) {
+        return new IO<>(delegate.flatMap(f.andThen(io -> io.delegate)::apply));
+    }
+
+    // TODO do we want a separate java Fiber?
+    public IO<Void, Fiber<E, A>> fork() {
+        return new IO<>(delegate.fork()).leftMap(nothing -> null);
+    }
+
+    public IO<Void, Fiber<E, A>> fork0(Function<ExitResult.Cause<Object>, IO<Void, Void>> f) {
+        return new IO<>(
+                delegate.fork0(f.andThen(io ->
+                        io.delegate.<Object, BoxedUnit>bimap(
+                                _void -> _void,
+                                _void -> BoxedUnit.UNIT
+                        ))::apply
+                )
+        ).bimap(nothing -> null, nothing -> null);
+    }
+
+    public <E1 extends E, B, C> IO<E1, C> parWith(IO<E1, B> that, BiFunction<A, B, C> f) {
+        return new IO<>(delegate.parWith(that.delegate, f::apply));
+    }
+
+    public <E1 extends E, B> IO<E1, scala.Tuple2<A, B>> par(IO<E1, B> that) {
```

Comment:
Is `scala.Tuple2` convenient to use from Java? If not, we should define our own.

Reply:
I think maybe we could introduce `Tuple2` from Vavr?

Reply:
The answer is no, Because I write Java every day.